File: src/schemas/crypto.ts
import { z } from "zod";
import { datasetEnvelope, utcDateTime } from "./envelope";

export const CryptoSentiment = z.enum(["risk_on", "risk_off", "neutral"]);

export const CryptoAsset = z
  .object({
    symbol: z.string().min(1),
    name: z.string().min(1),
    price: z.number().finite(),
    change_1d: z.number().finite().nullable(),
    change_1w: z.number().finite().nullable(),
    change_1m: z.number().finite().nullable(),
    market_cap: z.number().finite().nullable(),
    volume_24h: z.number().finite().nullable(),
    source: z.string().min(1),
    updated_at: utcDateTime,
  })
  .strict();

export const CryptoDataset = z
  .object({
    assets: z.array(CryptoAsset),
    btc_dominance: z.number().finite().min(0).max(1).nullable(),
    stablecoin_mcap: z.number().finite().nullable(),
    market_cap_total: z.number().finite().nullable(),
    sentiment: CryptoSentiment.nullable(),
  })
  .strict();

export const CryptoEnvelope = datasetEnvelope(CryptoDataset);

export type CryptoSentiment = z.infer<typeof CryptoSentiment>;
export type CryptoAsset = z.infer<typeof CryptoAsset>;
export type CryptoDataset = z.infer<typeof CryptoDataset>;
export type CryptoEnvelope = z.infer<typeof CryptoEnvelope>;

File: pipeline/schemas/crypto.py
"""加密资产数据集契约（架构 §3.2 cross_asset / 评审 §3 数据源矩阵 #7）。

T03 CoinGeckoCollector 负责填充；本模块只定义契约。
"""

from __future__ import annotations

from typing import Literal

from pydantic import Field

from .envelope import BaseEnvelope, ContractModel, UTCDateTime

CryptoSentiment = Literal["risk_on", "risk_off", "neutral"]


class CryptoAsset(ContractModel):
    symbol: str = Field(min_length=1, description="如 BTC / ETH / SOL")
    name: str = Field(min_length=1)
    price: float
    change_1d: float | None = None
    change_1w: float | None = None
    change_1m: float | None = None
    market_cap: float | None = None
    volume_24h: float | None = None
    source: str = Field(min_length=1)
    updated_at: UTCDateTime


class CryptoDataset(ContractModel):
    """crypto.json payload。"""

    assets: list[CryptoAsset] = Field(default_factory=list)
    btc_dominance: float | None = Field(default=None, ge=0.0, le=1.0)
    stablecoin_mcap: float | None = None
    market_cap_total: float | None = None
    sentiment: CryptoSentiment | None = None


class CryptoEnvelope(BaseEnvelope):
    payload: CryptoDataset
